### PR TITLE
🐛 Source Gitlab: handle expired `access_token` while checking the connection

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving: 
- https://github.com/airbytehq/oncall/issues/2928

## How
- Added intermediate `_try_refresh_access_token` and `_handle_expired_access_token_error` methods to catch and handle the error
- covered with unit test

## 🚨 User Impact 🚨
Not a breaking change.

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`7e6fe06796`](https://github.com/airbytehq/airbyte/pull/31210/commits/7e6fe06796e4f39255fe955cf04c916ca91b386d)) - ❌

⏲️  Total pipeline duration: 07mn46s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ❌        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1696937664/7e6fe06796e4f39255fe955cf04c916ca91b386d/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:7e6fe06796e4f39255fe955cf04c916ca91b386d)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py [hunk 1]
```diff
@@ -140,7 +141,35 @@ def _get_group_list(self, config: MutableMapping[str, Any]) -> List[str]:
...
+        # get currernt authenticator
+        authenticator: Union[SingleUseRefreshTokenOauth2Authenticator, TokenAuthenticator] = self.__auth_params.get("authenticator")
+        try:
+            creds = authenticator.refresh_access_token()
+            # update the actual config values
+            config["credentials"]["access_token"] = creds[0]
+            config["credentials"]["refresh_token"] = creds[3]
+            config["credentials"]["token_expiry_date"] = authenticator.get_new_token_expiry_date(creds[1], creds[2]).to_rfc3339_string()
+            # update the config
+            emit_configuration_as_airbyte_control_message(config)
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:158]: isn't it the same what's implemented in the [Airbyte CDK](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py#L224)?

Comment (bazarnov) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:158]: The  CDK version of this now doesn't attempt to refresh revoked or invalid `access_token` it doesn't refresh the `access_token` until the time comes for it to be expired. We explicitly trigger the `refresh`  procedure,

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`0ad3b1a8da`](https://github.com/airbytehq/airbyte/pull/31210/commits/0ad3b1a8da441839cc2ecadd179b0ba1b7c15029)) - ❌

⏲️  Total pipeline duration: 08mn01s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ❌        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1696939168/0ad3b1a8da441839cc2ecadd179b0ba1b7c15029/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:0ad3b1a8da441839cc2ecadd179b0ba1b7c15029)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py (hunk 1 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:158]: Thanks for the clarification. Can access tokens get revoked without revoking refresh tokens?
l think [this](https://github.com/airbytehq/airbyte/pull/31207) update should be enough, however other reviewers might have different thoughts

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`5ac17eaf1d`](https://github.com/airbytehq/airbyte/pull/31210/commits/5ac17eaf1dcabbbb9369a8f86f2cf919207892db)) - ❌

⏲️  Total pipeline duration: 07mn46s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ❌        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1696941848/5ac17eaf1dcabbbb9369a8f86f2cf919207892db/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:5ac17eaf1dcabbbb9369a8f86f2cf919207892db)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`fb76b03cab`](https://github.com/airbytehq/airbyte/pull/31210/commits/fb76b03cab55316e118334f8abe4ffffbf7cc40b)) - ❌

⏲️  Total pipeline duration: 07mn49s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1696947245/fb76b03cab55316e118334f8abe4ffffbf7cc40b/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:fb76b03cab55316e118334f8abe4ffffbf7cc40b)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`fb76b03cab`](https://github.com/airbytehq/airbyte/pull/31210/commits/fb76b03cab55316e118334f8abe4ffffbf7cc40b)) - ✅

⏲️  Total pipeline duration: 13mn40s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1696958824/fb76b03cab55316e118334f8abe4ffffbf7cc40b/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:fb76b03cab55316e118334f8abe4ffffbf7cc40b)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`3176f72f80`](https://github.com/airbytehq/airbyte/pull/31210/commits/3176f72f802329809a701ac66d3f3eaaa3119d71)) - ✅

⏲️  Total pipeline duration: 11mn18s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1697101592/3176f72f802329809a701ac66d3f3eaaa3119d71/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3176f72f802329809a701ac66d3f3eaaa3119d71)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`b8e498adbe`](https://github.com/airbytehq/airbyte/pull/31210/commits/b8e498adbe8f5877a46d4ff895c68260f2817e6d)) - ✅

⏲️  Total pipeline duration: 10mn51s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1697105079/b8e498adbe8f5877a46d4ff895c68260f2817e6d/source-gitlab/1.7.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:b8e498adbe8f5877a46d4ff895c68260f2817e6d)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`63017a8394`](https://github.com/airbytehq/airbyte/pull/31210/commits/63017a83949542521a5d1b0c3ca29bea6b48e8f7)) - ✅

⏲️  Total pipeline duration: 13mn00s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/baz_source-gitlab_handle-401-for-expired-access-token/1697197769/63017a83949542521a5d1b0c3ca29bea6b48e8f7/source-gitlab/1.7.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:63017a83949542521a5d1b0c3ca29bea6b48e8f7)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py [hunk 2]
```diff
@@ -154,8 +184,16 @@ def check_connection(self, logger, config) -> Tuple[bool, any]:
...
                 return True, None
             return True, None  # in case there's no projects
+        except HTTPError as http_error:
+            if config["credentials"]["auth_type"] == "oauth2.0":
+                if http_error.response.status_code == 401:
+                    return self._handle_expired_access_token_error(logger, config)
+                elif http_error.response.status_code == 500:
+                    return False, f"Unable to connect to Gitlab API with the provided credentials - {repr(http_error)}"
+            else:
+                return False, f"Unable to connect to Gitlab API with the provided Private Access Token - {repr(http_error)}"
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:194]: The only known is `500`, which is an `Internal Server Error`, while `check connection`. Not sure what to fix here, except for retry, but if the creds are bad - the retry will not be effective. These were before this fix, I've just moved them in the right way so we can handle all the present scenarios.

**Review by darynaishchenko — APPROVED** [review]

**Pull request merged**